crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,278 @@
+{
+    int i;
+
+    if (sk == NULL)
+        return 0;
+
+    if (certs == NULL)
+        return 1;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        X509 *cert = sk_X509_value(certs, i);
+        if (!no_self_signed || X509_check_issued(cert, cert) != X509_V_OK) {
+            if (!OSSL_CMP_sk_X509_add1_cert(sk, cert, no_duplicates, 0))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+int OSSL_CMP_X509_STORE_add1_certs(X509_STORE *store, STACK_OF(X509) *certs,
+                                   int only_self_signed)
+{
+    int i;
+
+    if (store == NULL)
+        return 0;
+
+    if (certs == NULL)
+        return 1;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        X509 *cert = sk_X509_value(certs, i);
+        if (!only_self_signed || X509_check_issued(cert, cert) == X509_V_OK)
+            if (!X509_STORE_add_cert(store, cert)) /* ups cert ref counter */
+                return 0;
+    }
+    return 1;
+}
+
+STACK_OF(X509) *OSSL_CMP_X509_STORE_get1_certs(X509_STORE *store)
+{
+    int i;
+    STACK_OF(X509) *sk;
+    STACK_OF(X509_OBJECT) *objs;
+
+    if (store == NULL)
+        return NULL;
+    if ((sk = sk_X509_new_null()) == NULL)
+        return NULL;
+    objs = X509_STORE_get0_objects(store);
+    for (i = 0; i < sk_X509_OBJECT_num(objs); i++) {
+        X509 *cert = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(objs, i));
+        if (cert != NULL) {
+            if (!sk_X509_push(sk, cert)) {
+                sk_X509_pop_free(sk, X509_free);
+                return NULL;
+            }
+            X509_up_ref(cert);
+        }
+    }
+    return sk;
+}
+
+X509_EXTENSIONS *CMP_X509_EXTENSIONS_dup(const X509_EXTENSIONS *exts)
+{
+    if (exts == NULL)
+        return NULL;
+    return sk_X509_EXTENSION_deep_copy(exts, X509_EXTENSION_dup,
+                                       X509_EXTENSION_free);
+}
+
+int CMP_ASN1_OCTET_STRING_set1(ASN1_OCTET_STRING **tgt,
+                               const ASN1_OCTET_STRING *src)
+{
+    if (tgt == NULL) {
+        CMPerr(CMP_F_CMP_ASN1_OCTET_STRING_SET1, CMP_R_NULL_ARGUMENT);
+        goto err;
+    }
+    if (*tgt == src) /* self-assignment */
+        return 1;
+    ASN1_OCTET_STRING_free(*tgt);
+
+    if (src != NULL) {
+        if ((*tgt = ASN1_OCTET_STRING_dup(src)) == NULL) {
+            CMPerr(CMP_F_CMP_ASN1_OCTET_STRING_SET1, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+    } else {
+        *tgt = NULL;
+    }
+
+    return 1;
+ err:
+    return 0;
+}
+
+int CMP_ASN1_OCTET_STRING_set1_bytes(ASN1_OCTET_STRING **tgt,
+                                     const unsigned char *bytes, size_t len)
+{
+    ASN1_OCTET_STRING *new = NULL;
+    int res = 0;
+
+    if (tgt == NULL) {
+        CMPerr(CMP_F_CMP_ASN1_OCTET_STRING_SET1_BYTES, CMP_R_NULL_ARGUMENT);
+        goto err;
+    }
+
+    if (bytes != NULL) {
+        if ((new = ASN1_OCTET_STRING_new()) == NULL
+                || !(ASN1_OCTET_STRING_set(new, bytes, (int)len))) {
```

Comment:
a prior `size_t -> int` owerflow test would be better IMHO.

Reply:
Here I changed the `len` parameter to `int`. Did it also in 2 calling functions.